oak/server/oak_node.cc

```diff
@@ -186,123 +141,119 @@ OakNode::OakNode(const std::string& node_id, const std::string& module)
-
   oak_module->AppendFuncExport(
-      "print",
+      "open_channel",
       wabt::interp::FuncSignature(std::vector<wabt::Type>{wabt::Type::I32, wabt::Type::I32},
-                                  std::vector<wabt::Type>{}),
-      PrintString(env));
-  oak_module->AppendFuncExport(
-      "get_time",
-      wabt::interp::FuncSignature(std::vector<wabt::Type>{},
                                   std::vector<wabt::Type>{wabt::Type::I64}),
-      OakGetTime(env));
-
-  oak_module->AppendFuncExport(
-      "read_method_name",
-      wabt::interp::FuncSignature(std::vector<wabt::Type>{wabt::Type::I32, wabt::Type::I32},
-                                  std::vector<wabt::Type>{wabt::Type::I32}),
-      this->OakReadMethodName(env));
+      this->OakOpenChannel(env));
   oak_module->AppendFuncExport(
-      "read",
-      wabt::interp::FuncSignature(std::vector<wabt::Type>{wabt::Type::I32, wabt::Type::I32},
-                                  std::vector<wabt::Type>{wabt::Type::I32}),
-      this->OakRead(env));
+      "read_channel",
+      wabt::interp::FuncSignature(
+          std::vector<wabt::Type>{wabt::Type::I64, wabt::Type::I32, wabt::Type::I32},
+          std::vector<wabt::Type>{wabt::Type::I32}),
+      this->OakReadChannel(env));
   oak_module->AppendFuncExport(
-      "write",
-      wabt::interp::FuncSignature(std::vector<wabt::Type>{wabt::Type::I32, wabt::Type::I32},
-                                  std::vector<wabt::Type>{wabt::Type::I32}),
-      this->OakWrite(env));
+      "write_channel",
+      wabt::interp::FuncSignature(
+          std::vector<wabt::Type>{wabt::Type::I64, wabt::Type::I32, wabt::Type::I32},
+          std::vector<wabt::Type>{wabt::Type::I32}),
+      this->OakWriteChannel(env));
+  // TODO: Implement and register close_channel function.
 }
 
-wabt::interp::HostFunc::Callback OakNode::OakReadMethodName(wabt::interp::Environment* env) {
+ChannelId OakNode::NewChannelId() {
+  // TODO: Generate random channel ID.
+  channel_id_ += 1;
+  return channel_id_;
+}
+
+wabt::interp::HostFunc::Callback OakNode::OakOpenChannel(wabt::interp::Environment* env) {
   return [this, env](const wabt::interp::HostFunc* func, const wabt::interp::FuncSignature* sig,
                      const wabt::interp::TypedValues& args, wabt::interp::TypedValues& results) {
-    LOG(INFO) << "Called host function: " << func->module_name << "." << func->field_name;
-    for (auto const& arg : args) {
-      LOG(INFO) << "Arg: " << wabt::interp::TypedValueToString(arg);
+    LogHostFunctionCall(func, args);
+
+    std::string name = ReadString(env, args[0].get_i32(), args[1].get_i32());
+    LOG(INFO) << "open_channel(" << name << ")";
+
+    ChannelId channel_id = NewChannelId();
+    std::unique_ptr<Channel> channel;
+    if (name == "grpc_method_name") {
+      absl::Span<const char> grpc_method_name_span = absl::MakeConstSpan(grpc_method_name_);
+      channel = absl::make_unique<BufferChannel>(grpc_method_name_span, nullptr);
```

Comment:
Could eliminate the temporary and just inline absl::MakeConstSpan here.

Reply:
Not relevant any more.